### PR TITLE
 Handle InstallationReport in DeviceManifest 

### PR DESCRIPTION
**Pull request by jerrytrieu**
Built on #153

### CODE REVIEW

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 1]
```diff
@@ -76,4 +90,46 @@ class DeviceManifestUpdate(afterUpdate: AfterDeviceManifestUpdate,
...
+      } else {
+        successInstallationResult
+      }
+
+      if (ecuResults.isEmpty && enforceExplicit) {
+        None
+      } else {
+       val result = Some(DeviceInstallationReport(
+          namespace, deviceId, correlationId, result, ecuResults, None, Instant.now))
+      _log.error(s"Old DeviceInstallationReport ${result}")
```
Comment (jochenschneider) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: indentation

File: src/main/scala/com/advancedtelematic/director/manifest/AfterUpdate.scala [hunk 2]
```diff
@@ -1,71 +1,66 @@
...
+import com.advancedtelematic.libtuf.data.TufDataType.{TargetFilename, ValidTargetFilename}
 
 import java.time.Instant
-import scala.async.Async._
 import scala.concurrent.{ExecutionContext, Future}
 import slick.driver.MySQLDriver.api._
 
 
 class AfterDeviceManifestUpdate()
                                (implicit db: Database, ec: ExecutionContext,
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/manifest/AfterUpdate.scala]: Should rename the file to `AfterDeviceManifestUpdate`?

File: src/main/scala/com/advancedtelematic/director/manifest/AfterUpdate.scala [hunk 3]
```diff
@@ -1,71 +1,66 @@
...
+            updateTarget.correlationId.get,
+            failureInstallationResult,
+            Map(), None, Instant.now))
+      }
+    } yield ()
 
-    }
+  private def toFailedTargets(report: DeviceInstallationReport) : Map[EcuSerial, TargetFilename] = {
+    report.ecuReports
+      .filter{ case (k, v) => !v.result.success }
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/manifest/AfterUpdate.scala]: Maybe clearer: `.filterNot(_._2.result.success)` ?

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 4]
```diff
@@ -28,22 +36,25 @@ class DeviceManifestUpdate(afterUpdate: AfterDeviceManifestUpdate,
...
+        _log.error(s"UpdateNotCompleted $updateTarget")
+        val f = for {
+          correlationId <- updateTarget.correlationId
+          ireport <- toDeviceInstallationReport(namespace, device, deviceManifest, correlationId)
+          report = ireport if (!ireport.result.success)
+        } yield afterUpdate.clearUpdate(report)
+        f.getOrElse(Future.successful(()))
       case DeviceUpdateResult.UpdateSuccessful(updateTarget) =>
-        afterUpdate.successMultiTargetUpdate(namespace, device, updateTarget, ecuResults)
+        _log.error(s"UpdateSuccessful $updateTarget")
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: Why log error when UpdateSuccessful?

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 5]
```diff
@@ -78,4 +90,46 @@ class DeviceManifestUpdate(afterUpdate: AfterDeviceManifestUpdate,
...
+      val report = ireport.report
+      val ecuResults = report.items
+        .filter(item => ecuImages.contains(item.ecu))
+        .map { item =>
+          item.ecu -> EcuInstallationReport(item.result, Seq(ecuImages.get(item.ecu).get.toString), None)
+        }.toMap
+
+      val result = DeviceInstallationReport(
+        namespace, deviceId, report.correlation_id, report.result, ecuResults, None, Instant.now)
+      _log.error(s"New DeviceInstallationReport ${result}")
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: Also here, should it be error?

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 6]
```diff
@@ -78,4 +90,46 @@ class DeviceManifestUpdate(afterUpdate: AfterDeviceManifestUpdate,
...
+      result
+
+    }.orElse {
+
+      // Support legacy `custom.operation_result`
+      val ecuResults = toEcuInstallationReports(deviceManifest.ecu_manifests)
+      val result = if (ecuResults.find(!_._2.result.success).isDefined) {
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: `if (ecuResults.exists(!_._2.result.success))`

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 7]
```diff
@@ -76,4 +91,46 @@ class DeviceManifestUpdate(afterUpdate: AfterDeviceManifestUpdate,
...
+    val ecuImages = deviceManifest.ecu_manifests.map { ecuManifest =>
+      ecuManifest.ecu_serial -> ecuManifest.installed_image.filepath
+    }.toMap
+
+    deviceManifest.installation_report.map { ireport =>
+      val report = ireport.report
+      val ecuResults = report.items
+        .filter(item => ecuImages.contains(item.ecu))
+        .map { item =>
+          item.ecu -> EcuInstallationReport(item.result, Seq(ecuImages.get(item.ecu).get.toString), None)
```
Comment (houcros) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: `Seq(ecuImages(item.ecu).toString)`

**Pull request merged**